rs/src/connections/host_relay.rs

```diff
@@ -0,0 +1,1085 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+use std::{
+    collections::HashMap,
+    io,
+    net::{Ipv4Addr, SocketAddr, SocketAddrV4},
+    pin::Pin,
+    sync::Arc,
+    task::Poll,
+    time::Duration,
+};
+
+use crate::{
+    contracts::{TunnelConnectionMode, TunnelEndpoint, TunnelPort, TunnelRelayTunnelEndpoint},
+    management::{
+        Authorization, HttpError, TunnelLocator, TunnelManagementClient, TunnelRequestOptions,
+        NO_REQUEST_OPTIONS,
+    },
+};
+use futures::{FutureExt, TryFutureExt};
+use russh::{server::Server as ServerTrait, CryptoVec};
+use tokio::{
+    io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt},
+    net::TcpStream,
+    sync::{mpsc, oneshot, watch},
+    task::JoinHandle,
+};
+use tokio_tungstenite::{connect_async, MaybeTlsStream, WebSocketStream};
+use uuid::Uuid;
+
+use super::{
+    errors::TunnelError,
+    ws::{build_websocket_request, AsyncRWWebSocket},
+};
+
+/// Mapping of port numbers to senders to which new port connections should be
+/// sent. Shared by the host relay to each connected session.
+type PortMap = HashMap<u32, mpsc::UnboundedSender<ForwardedPortConnection>>;
+
+/// The HostRelay can host connections via the tunneling service. After
+/// creating it, you will generally want to run `connect()` to create a new
+/// a new connection.
+///
+/// Note that, while ports can be added and remove dynamically from running
+/// tunnels via the appropriate methods on the HostRelay, no ports will be
+/// hosted until those methods are called.
+pub struct HostRelay {
```

Comment:
I'm not sure about this name. In the tunneling terminology we use, the "relay" is the cloud service that relays data between host and client.

In other SDKs we call this thing `TunnelRelayTunnelHost`, where "relay" is describing the specific type of tunnel host. (Just `RelayTunnelHost` might be simpler, but there used to be another host type that used Live Share relays; that's obsolete now.) In the future there will be other host classes, like a `LocalNetworkTunnelHost` that accepts direct connections on a local network without going through a relay.